src/spn/algorithms/MEU.py

```diff
@@ -0,0 +1,162 @@
+"""
+Created on March 28, 2019
+@author: Hari Teja Tatavarti
+
+"""
+
+from spn.algorithms.MPE import get_node_funtions
+from spn.algorithms.Inference import  likelihood, max_likelihood, log_likelihood, sum_likelihood, prod_likelihood
+from spn.algorithms.Validity import is_valid
+from spn.structure.Base import get_nodes_by_type, Max, Leaf, Sum, Product, get_topological_order_layers
+
+import numpy as np
+from collections import defaultdict
+import collections
+
+
+def merge_input_vals(l):
+    return np.concatenate(l)
+
+def meu_max(node, parent_result, data=None, lls_per_node=None, rand_gen=None):
+    if len(parent_result) == 0:
+        return None
+
+    parent_result = merge_input_vals(parent_result)
+
+    w_children_log_probs = np.zeros((len(parent_result), len(node.dec_values)))
+    for i, c in enumerate(node.children):
+        w_children_log_probs[:, i] = lls_per_node[parent_result, c.id]
+
+    max_child_branches = np.argmax(w_children_log_probs, axis=1)
+
+    children_row_ids = {}
+
+    for i, c in enumerate(node.children):
+        children_row_ids[c] = parent_result[max_child_branches == i]
+    # print("node and max_child_branches", (node.var_name, max_child_branches))
+    ## decision values at each node
+
+    decision_values = {}
+    decision_values[node.feature_name] = np.column_stack((parent_result, max_child_branches))
+
+    # print("w_children_log_probs",w_children_log_probs)
+    return children_row_ids, decision_values
+
+
+
+node_functions =  get_node_funtions()
+_node_top_down_meu= node_functions[0].copy()
+_node_bottom_up_meu = node_functions[1].copy()
+_node_top_down_meu.update({Max:meu_max})
+_node_bottom_up_meu.update({Sum: sum_likelihood, Product: prod_likelihood, Max:max_likelihood})
+
+
+
+
+def meu(node, input_data, node_top_down_meu=_node_top_down_meu, node_bottom_up_meu=_node_bottom_up_meu, in_place=False):
+    valid, err = is_valid(node)
+    assert valid, err
+    if in_place:
```

Comment:
Sorry, I'm a little confused as to what the `in_place` is supposed to do here? why not just explicitly cast to an `np.array` and have an assert on the top to ensure that it is in fact a list, so we don't get any ugly stack traces of converting to an `np.array` breaking (or `np.ndarray`)

Reply:
Again, I left `in_place` in place to be consistent with `mpe` function in MPE.py. And yeah, having an assert and explicitly casting to `np.array` seems to be a good alternative.